rbd/migration.go

```diff
@@ -0,0 +1,240 @@
+
+// #cgo LDFLAGS: -lrbd
+// /* force XSI-complaint strerror_r() */
+// #define _POSIX_C_SOURCE 200112L
+// #undef _GNU_SOURCE
+// #include <errno.h>
+// #include <stdlib.h>
+// #include <rados/librados.h>
+// #include <rbd/librbd.h>
+import "C"
+import (
+	"unsafe"
+
+	"github.com/ceph/go-ceph/rados"
+)
+
+// cMigrationData contains the data needed by the migration functions
+type cMigrationData struct {
+	sourceSpec      *C.char
+	sourceImageName *C.char
+	destImageName   *C.char
+	free            func()
+}
+
+// MigrationOptions interface is used to encapsulate the different migrations
+// options and enable converting them from go to C structures.
+type MigrationOptions struct {
+	SourceSpec  string
+	SourceImage string
+	DestImage   string
+}
+
+// allocateMigrationOptions interface is used to encapsulate the different
+// migration options and enables converting them from Go to C structures.
+func (opts MigrationOptions) allocateMigrationOptions() cMigrationData {
+	var retData cMigrationData
+	//CBytes allocates memory which we'll free by calling cOptsFree()
+	retData.sourceSpec = (*C.char)(C.CBytes([]byte(opts.SourceSpec)))
+	retData.sourceImageName = (*C.char)(C.CBytes([]byte(opts.SourceImage)))
+	retData.destImageName = (*C.char)(C.CBytes([]byte(opts.DestImage)))
+	retData.free = func() {
+		C.free(unsafe.Pointer(retData.sourceSpec))
+		C.free(unsafe.Pointer(retData.sourceImageName))
+		C.free(unsafe.Pointer(retData.destImageName))
+	}
+	return retData
+}
+
+// MigrationPrepare prepares a migration
+//
+// Implements:
+//  int rbd_migration_prepare(rados_ioctx_t ioctx,
+// 							  const char *image_name,
+// 							  rados_ioctx_t dest_ioctx,
+// 							  const char *dest_image_name,
+// 							  rbd_image_options_t opts);
+//
+func MigrationPrepare(ioctx *rados.IOContext, destIoctx *rados.IOContext, opts MigrationOptions, rio *ImageOptions) error {
+	migrationOpts := opts.allocateMigrationOptions()
+	defer migrationOpts.free()
+
+	ret := C.rbd_migration_prepare(
+		cephIoctx(ioctx),
+		migrationOpts.sourceImageName,
+		cephIoctx(destIoctx),
+		migrationOpts.destImageName,
+		C.rbd_image_options_t(rio.options))
+
+	return getError(ret)
+}
+
+// MigrationPrepareImport prepares a migration for import
+// from a specified source to a specified image
+//
+// Implements:
+//  int rbd_migration_prepare_import(const char *source_spec,
+//									 rados_ioctx_t dest_ioctx,
+//     								 const char *dest_image_name,
+//									 rbd_image_options_t opts);
+//
+func MigrationPrepareImport(ioctx *rados.IOContext, opts MigrationOptions, rio *ImageOptions) error {
+	migrationOpts := opts.allocateMigrationOptions()
+	defer migrationOpts.free()
+
+	ret := C.rbd_migration_prepare_import(
+		migrationOpts.sourceSpec,
+		cephIoctx(ioctx),
+		migrationOpts.destImageName,
+		C.rbd_image_options_t(rio.options))
+
+	return getError(ret)
+}
+
+// MigrationExecute executes a migration starting
+// the full migration of data into image from source
```

Comment:
Style nitpick. The first line of a doc comment should be a full english sentence that starts with the name of the item being documented and ending with a period.

Reply:
Changed